### PR TITLE
[precompile] Add torch.compiler.precompile (capture + eager backend)

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #190226
* #189622
* #189165
* #189621
* #189629
* __->__ #189620

> **Split for PR size / reviewability.** This change was split into an
> implementation PR and a tests PR so each stays under PyTorch's ~2000-line PR
> sanity-check (the CI lint that flags oversized diffs). **This PR is the
> implementation only**; its test suite lives in the stacked follow-up
> **#189629** ([precompile] Tests for torch.compiler.precompile). The two are
> meant to land together.

Adds torch.compiler.precompile, an ahead-of-time precompile API that captures a
whole computation with make_fx and lowers it to a self-contained, runnable Python
source string plus an integrity-tagged cache:

    python_code, cache = torch.compiler.precompile(fn, model, *example_inputs)
    f_c = torch.compiler.precompile.load(python_code, cache)
    f_c(model, *example_inputs)   # pass the model again; grads land on .grad

python_code is the single source of truth for the calling convention and runs on
its own; no model weights are baked in -- params/buffers are lifted to graph inputs
and re-extracted from the runtime model each call. The reloaded driver validates
every call (structure, shape, dtype, device) and fails with PrecompileError rather
than silently mis-computing.

This first PR implements the backend="eager" path (the captured ATen graph is
inlined and run as-is) plus the full capture, calling-convention metadata, cache
envelope, and public API surface. The inductor backend and dynamic shapes follow
in later PRs in this stack; backend="inductor" raises for now.

Test Plan:

Unit tests for this implementation live in the stacked #189629:

```
python test/test_precompile.py
```

Public-bindings checks (this PR adds the torch.compiler.precompile public API):

```
python test/test_public_bindings.py \
    TestPublicBindings.test_correct_module_names \
    TestPublicBindings.test_no_new_reexport_callables
```

Authored with an AI assistant.

### CODE REVIEW

File: torch/_precompile.py [hunk 1]
```diff
@@ -0,0 +1,1463 @@
...
+                    full.append(runtime_inputs[ui])
+                    ui += 1
+            result = fn(*full)
+            # Harvest parameter gradients produced by any backward in fn.
+            param_proxies = pb[:num_params]
+            harvested = [p.grad for p in param_proxies]
+            # Buffers are not harvested (only params get scattered grads). A registered
+            # buffer with requires_grad=True that received a gradient would be silently
+            # dropped, so reject it -- a cheaply-knowable invariant-5 violation.
+            if any(getattr(b, "grad", None) is not None for b in pb[num_params:]):
```
Comment (aorenste) [inline torch/_precompile.py:481]: [blocker] A user-input leaf that requires grad and receives a gradient during the traced backward is neither harvested nor rejected here -- it is silently dropped. Confirmed empirically: for `lambda model, x: model(x).sum().backward()` with `x.requires_grad_(True)`, eager fills `x.grad`, but the reloaded artifact leaves the runtime input's `.grad` at None and capture raises nothing. That breaks the fail-loud contract the adjacent buffer check upholds. Mirror it for user inputs (`flat[num_pb:]` are the user leaves in scope here):

```python
if any(getattr(t, "grad", None) is not None for t in flat[num_pb:]):
    raise PrecompileError(
        "precompile: a user input received a gradient; precompile only "
        "harvests gradients for parameters, so an input gradient would be "
        "silently dropped."
    )
```

(Or harvest input grads as extra outputs and scatter them back -- but rejecting keeps parity with the buffer case.)

File: torch/_precompile.py [hunk 2]
```diff
@@ -0,0 +1,1463 @@
...
+        """Build a runnable from load()'s reconstructed forward.
+
+        load() does not re-run capture/_compile, so reuse ``__init__`` for all the
+        defaults (the single definition of this object's state) and override only the
+        calling-convention fields recovered from python_code plus the reconstructed
+        forward. The capture-only fields (``_fn``, ``_gm``, ``_param_names``, ...)
+        stay at their ``__init__`` defaults; inspect the artifact via python_code.
+        """
+        obj = cls(None, backend=backend)  # type: ignore[arg-type]
+        obj._module_positions = module_positions
```
Comment (aorenste) [inline torch/_precompile.py]: [cleanup] These four fields are dead on a loaded object: `__call__` only delegates to `_loaded_forward` (the inlined driver, which carries its own MODULE_POSITIONS / NUM_POSITIONAL_ARGS / GRAD_PARAM_INDICES / OUT_SPEC), and the only readers (`to_python_code` / `to_cache_bytes`) raise for loaded objects. `load()`'s extraction of these and these assignments can be dropped (keep `_parse_artifact_metadata` for its validation of `python_code`); `_from_loaded` then needs only `forward` and `backend`.

File: torch/_precompile.py [hunk 3]
```diff
@@ -0,0 +1,1463 @@
...
+    """The error type raised by ``torch.compiler.precompile`` and its artifacts.
+
+    Raised when capture, ``load``, or a runtime call violates the precompile contract --
+    e.g. a tensor baked as a constant (invariant 1), an unsupported / effectful op, or a
+    runtime input whose shape / dtype / device differs from the example (invariants 3 and
+    6). See Note [precompile programming model] in this module for the full contract.
+    """
+
+
+def _dense_shape(t: Any) -> tuple[int, ...] | None:
```
Comment (aorenste) [inline torch/_precompile.py]: nit (here and all the `Any` types below)
```suggestion
def _dense_shape(t: object) -> tuple[int, ...] | None:
```

File: torch/_precompile.py [hunk 4]
```diff
@@ -0,0 +1,1463 @@
...
+
+    Tensor subclasses (e.g. DTensor) carry an outer shape that is not a plain dense
+    shape; record ``None`` and skip them in the shape check.
+    """
+    if isinstance(t, torch.Tensor) and not is_traceable_wrapper_subclass(t):
+        return tuple(t.shape)
+    return None
+
+
+def _dense_dtype(t: Any) -> Any:
```
Comment (aorenste) [inline torch/_precompile.py]:
```suggestion
def _dense_dtype(t: object) -> torch.dtype | None:
```

File: torch/_precompile.py [hunk 5]
```diff
@@ -0,0 +1,1463 @@
...
+    metadata as a literal and compares cleanly against ``str(t.dtype)`` at runtime;
+    mirrors the _dense_shape convention (None for non-tensor / subclass leaves). The
+    graph is specialized to the example dtype (invariant 6).
+    """
+    if isinstance(t, torch.Tensor) and not is_traceable_wrapper_subclass(t):
+        return str(t.dtype)
+    return None
+
+
+def _dense_device(t: Any) -> Any:
```
Comment (aorenste) [inline torch/_precompile.py]:
```suggestion
def _dense_device(t: object) -> torch.device | None:
```

File: torch/_precompile.py [hunk 6]
```diff
@@ -0,0 +1,1463 @@
...
+    is an input.
+
+    Every legitimate tensor in a non-strict capture is a placeholder (a lifted
+    parameter/buffer or user input) or the result of a ``call_function`` node.
+    A ``get_attr`` pointing at a tensor therefore means some tensor was closed
+    over (a global, captured local, or non-registered module attribute) and would
+    be baked into the graph as a constant, which we forbid.
+    """
+    offending = []
+    for node in gm.graph.nodes:
```
Comment (aorenste) [inline torch/_precompile.py]: [cleanup] The iterate-and-filter can be the more direct `gm.graph.find_nodes(op="get_attr")` (backed by an internal per-op lookup table), which states intent and skips the whole-graph scan:

```python
for node in gm.graph.find_nodes(op="get_attr"):
    attr = gm
    for part in node.target.split("."):
        attr = getattr(attr, part, None)
    if isinstance(attr, torch.Tensor):
        offending.append((node.target, tuple(attr.shape), str(attr.dtype)))
```

The dotted walk could also reuse `torch.fx.graph_module._get_attr(gm, node.target)`, but that is a private fx helper that raises on a missing attr rather than tolerating None, so `find_nodes` alone is the safe win.

File: torch/_precompile.py [hunk 7]
```diff
@@ -0,0 +1,1463 @@
...
+
+    They appear as ``get_attr`` nodes pointing at nested ``GraphModule`` submodules.
+    The eager backend inlines ``gm.code`` and cannot reach such submodules (they are
+    not on the standalone ``_GraphSelf`` holder), so the artifact would crash at
+    runtime. Fail at capture with a concrete reason instead, like ``_assert_supported``.
+    """
+    # Resolve the target the same way as _check_no_constant_tensors (dotted walk), so a
+    # nested-qualname subgraph is not silently missed.
+    offending = []
+    for node in gm.graph.nodes:
```
Comment (aorenste) [inline torch/_precompile.py]: [cleanup] Same iterate-and-filter as `_check_no_constant_tensors`: use `gm.graph.find_nodes(op="get_attr")` instead of scanning every node and filtering by op:

```python
for node in gm.graph.find_nodes(op="get_attr"):
    attr = gm
    for part in node.target.split("."):
        attr = getattr(attr, part, None)
    if isinstance(attr, torch.fx.GraphModule):
        offending.append(node.target)
```

File: torch/_precompile.py [hunk 8]
```diff
@@ -0,0 +1,1463 @@
...
+            "closed over by fn (a global or captured local) or a plain "
+            "(non-registered) module attribute, and also for a tensor literal "
+            "constructed inside fn (e.g. torch.tensor([...])). Offending constants "
+            f"(target, shape, dtype): {offending}. Fix by passing the tensor as an "
+            "explicit argument; for module state register it as a parameter/buffer, "
+            "and for a literal hoist it out of fn and pass it as an argument."
+        )
+
+
+def _assert_no_control_flow_subgraphs(gm: torch.fx.GraphModule) -> None:
```
Comment (aorenste) [inline torch/_precompile.py:291]: Maybe we need a helper to grab filtered get_attr nodes that this and _check_no_constant_tensors share?

File: torch/_precompile.py [hunk 9]
```diff
@@ -0,0 +1,1463 @@
...
+
+    INVARIANT: the all-modules' params then all-modules' buffers, dedup-by-id ordering
+    here is load-bearing and is reproduced VERBATIM by the embedded
+    ``_extract_param_buffers`` in _EAGER_DRIVER_SOURCE (the inlined eager load path), so
+    the two must stay in sync.
+    """
+    multi = len(mods) > 1
+
+    def _name(mi: int, n: str) -> str:
+        return f"m{mi}.{n}" if multi else n
```
Comment (aorenste) [inline torch/_precompile.py]: Uber nit: Slightly faster as:
```suggestion
    if len(mods) > 1:
        def _name(mi: int, n: str) -> str:
            return f"m{mi}.{n}"
    else:
        def _name(mi: int, n: str) -> str:
            return n
```
(could also make _name_multi and _name_single at toplevel and just assign to a local here...)

File: torch/_precompile.py [hunk 10]
```diff
@@ -0,0 +1,1463 @@
...
+    user-supplied modules, so the dense list lines up with the compiled graph.
+
+    INVARIANT: the all-modules' params then all-modules' buffers, dedup-by-id ordering
+    here is load-bearing and is reproduced VERBATIM by the embedded
+    ``_extract_param_buffers`` in _EAGER_DRIVER_SOURCE (the inlined eager load path), so
+    the two must stay in sync.
+    """
+    multi = len(mods) > 1
+
+    def _name(mi: int, n: str) -> str:
```
Comment (aorenste) [inline torch/_precompile.py]: Consider using a NewType to document the `mi` type (here and everywhere it's used).

File: torch/_precompile.py [hunk 11]
```diff
@@ -0,0 +1,1463 @@
...
+        raise PrecompileError(
+            "precompile cannot lower a captured control-flow subgraph (e.g. from "
+            f"torch.cond / torch.while_loop); not supported yet. Offending get_attr "
+            f"targets: {offending}."
+        )
+
+
+def _intern_param_buffers(
+    mods: list[torch.nn.Module],
+) -> tuple[list[Any], list[str], list[str], list[tuple[int, str, int]], int]:
```
Comment (aorenste) [inline torch/_precompile.py]:
```suggestion
) -> tuple[list[Tensor], list[str], list[str], list[tuple[int, str, int]], int]:
```

Way too many 'Any' types in this code. Anything that is really "anything" should be object but a lot of them are just lazy types. I'm going to stop pointing them out and just assume you're going to fix them...

File: torch/_precompile.py [hunk 12]
```diff
@@ -0,0 +1,1463 @@
...
+    buffer_dtypes = [str(t.dtype) for t in pb_flat[num_params:]]
+    param_devices = [str(t.device) for t in pb_flat[:num_params]]
+    buffer_devices = [str(t.device) for t in pb_flat[num_params:]]
+
+    user_flat, in_spec = pytree.tree_flatten(user_inputs)
+    flat_args = [*pb_flat, *user_flat]
+    # The REAL example tensors (params/buffers and user inputs). The saved-grad
+    # snapshot/clear/restore block below protects the real example model's .grad fields
+    # (those are what the user owns and what a backward in fn populates).
+    real_flat = [*pb_flat, *user_flat]
```
Comment (aorenste) [inline torch/_precompile.py]: nit:
```suggestion
    real_flat = list(flat_args)
```

**Review by aorenste — DISMISSED** [review]
(Reviewed by me, assisted by AI)

Note: I haven't finished reviewing this yet, but there's enough feedback here for a round. Requesting changes.

[cleanup] torch/_precompile.py has a very high comment-to-code ratio, and several points (invariants 2/3/6, "single source of truth", "integrity-tagged envelope", "no weights baked") are restated near-verbatim across the module docstring, Note [precompile programming model], the per-field comments in `PrecompiledModule.__init__` and `_build_metadata_section`, and the generated header. Repo style is "minimize comments; code should be self-documenting" -- consider keeping the canonical statement in the Note and trimming the per-field restatements. Lowest priority.

File: torch/_precompile.py [hunk 13]
```diff
@@ -0,0 +1,1407 @@
...
+    # invariant-1 guard with a misleading "tensor closed over by fn" error on the common
+    # warmup-step-then-precompile flow. Restored in finally; precompile does not mutate the
+    # user's example .grad (params/buffers AND user inputs). Snapshot the ORIGINAL .grad
+    # object (no clone) and restore that SAME object below, so grad IDENTITY is preserved --
+    # a caller holding a prior p.grad reference, or optimizer state keyed on grad identity,
+    # is not invalidated. Tracing runs on the real interned params, so a backward in fn DOES
+    # write .grad in place -- but onto a fresh grad object, since .grad was snapshotted and
+    # cleared to None just above; the finally-restore puts the snapshotted object back.
+    saved_grads = [
+        a.grad if isinstance(a, torch.Tensor) and a.grad is not None else None
```
Comment (aorenste) [inline torch/_precompile.py]: nit
```suggestion
        a.grad if isinstance(a, torch.Tensor) else None
```

File: torch/_precompile.py [hunk 14]
```diff
@@ -0,0 +1,1407 @@
...
+        user_input_devices=user_input_devices,
+    )
+
+
+class _Capture:
+    def __init__(
+        self,
+        gm: torch.fx.GraphModule,
+        flat_args: list[object],
+        module_positions: list[int],
```
Comment (aorenste) [inline torch/_precompile.py:564]: [cleanup] These calling-convention fields (module_positions, param_names, buffer_names, the *_shapes / *_dtypes / *_devices, grad_param_indices, user_input_*) are set once at capture and never mutated -- consider `tuple` over `list` to document that and guard against accidental mutation of the calling convention (this module leans hard on "load-bearing / must stay in sync" invariants). The inner shapes are already `tuple[int, ...]`, so `tuple[tuple[int, ...], ...]` is also more consistent. Safe to switch: nothing mutates them, `ast.literal_eval` round-trips the emitted `(...)` literals, and the driver already coerces via `list(PARAM_NAMES)` / `list(PARAM_SHAPES)`. Keep `flat_args` a `list` (it is the working set of live example tensors, not frozen metadata). If done, mirror it in `PrecompiledModule`'s fields and its `__init__` defaults (`[]` -> `()`). Low priority.

File: torch/_precompile.py [hunk 15]
```diff
@@ -0,0 +1,1407 @@
...
+    found: dict[str, object] = {}
+    try:
+        tree = ast.parse(python_code)
+    except SyntaxError as e:
+        raise PrecompileError(
+            "python_code is not valid Python; it does not look like a "
+            "torch.compiler.precompile artifact."
+        ) from e
+    for node in tree.body:
+        if not isinstance(node, ast.Assign) or len(node.targets) != 1:
```
Comment (aorenste) [inline torch/_precompile.py:712]: Intentionally silently dropping unknown lines? Sounds like a recipe for confusion at best. Should at least log.

File: torch/_precompile.py [hunk 16]
```diff
@@ -0,0 +1,1407 @@
...
+    parts.append("")
+    parts.extend(_build_metadata_section(compiled))
+    parts.append("# " + "=" * 70)
+    parts.append("# 3. Driver: run the inlined captured graph eagerly")
+    parts.append("# " + "=" * 70)
+    parts.append(_EAGER_DRIVER_SOURCE)
+    return "\n".join(parts)
+
+
+_EAGER_DRIVER_SOURCE = '''
```
Comment (aorenste) [inline torch/_precompile.py:799]: [maintainability / follow-up] The whole driver living in this string means our tooling (pyrefly, ruff, IDE nav/refactor) has zero visibility into ~160 lines of load-bearing runtime code -- hence the wall of `# noqa: F821`. We can author it as real code and still emit it as text.

Constraint to preserve: `python_code` must stay a self-contained, version-frozen artifact (behavior is hashed via `code_hash`, invariant 7), so the fix is NOT to import the driver from the artifact (that reintroduces torch-version skew). Keep emitting text; just stop hand-writing it as a string.

Two options:
- Lightweight: move the driver into a real module (e.g. `torch/_precompile_driver.py`) with the injected names (`PARAM_NAMES`, `call`, `_torch`, ...) declared under `if TYPE_CHECKING:` so pyrefly type-checks the bodies, then emit it via `inspect.getsource(...)` after a sentinel marker. Run the `getsource` LAZILY on the emit / `to_python_code` path only -- `load()` never runs it, so the "source stripped in prod / pyc-only" failure mode never applies.
- Robust: generate the `_EAGER_DRIVER_SOURCE` constant at build time via codegen (this repo is already codegen-heavy), so there is no runtime `inspect` at all.

Bonus either way: the driver becomes directly unit-testable (import and call `_extract_param_buffers`), and we can assert the sync invariant directly -- `_extract_param_buffers(mods)[1] == _intern_param_buffers(mods)[1]` -- instead of relying on the integration tests to catch drift.

Not for this PR -- capturing as a follow-up direction. (Minor precedent that the artifact isn't dogmatically zero-import: `_fail` already lazily imports `PrecompileError`; it's the behavior that must stay inlined/frozen.)

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

LGTM -- the earlier round is addressed. The remaining inline comments are optional nits and a follow-up direction, not blockers.

File: torch/_precompile.py (hunk 16 above)
Comment (bobrenjc93) [reply torch/_precompile.py:799]: will stack on a refactor at the top of the stack
